src/Components/test/testassets/TestServer/Infrastructure/InterruptibleWebSocketFeature.cs

```diff
@@ -0,0 +1,247 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Net.WebSockets;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.WebSockets;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace Components.TestServer.Infrastructure
+{
+    public class InterruptibleWebSocketFeature : IHttpWebSocketFeature
```

Comment:
Ok, this is really starting to see like a lot of bad duplication that we don't want

Reply:
This is the piece that implements the handshake. We do it this way so that we can get access to the websocket and throw whatever exception we want on top to cause issues on signalr that simulate an ungraceful disconnect.

We could have some class made public on the websockets front (UpgradeHandshake) and that would solve all our problems.
Either that or we reference the source or something.

Reply:
I'm going to keep this as is for now. There are three options we have here:
* Leave it like this. I don't think this code will change, so no problem.
* Make UpgradeHandshake public, then we just consume it.
* Have some sort of shared source arrangement.

For now I don't think this is important and can be kept here as a private implementation detail for the test.

Reply:
No, I don't think that's OK. This warrants a discussion with the people who are more familiar with this area.

Reply:
@Tratcher @anurse @BrennanConroy @jkotalik - what are some options here?

Reply:
What's the goal here? If you want to replicate the behavior of the actual server-side WebSocket, I think forcing it to throw an `IOException` isn't going to help since I don't believe the [server-side WebSocket ever does such a thing](https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs).

If you do need to wrap the WebSocket, couldn't you just wrap the existing feature and delegate to it, then wrap the socket it returns? That wouldn't need to duplicate the entire feature. Have `UseInterruptableWebSockets` put the real WebSockets middleware in first, then `InterruptableWebSocketMiddleware`. Then the middleware can capture the real `IHttpWebSocketFeature` and wrap that. You'd still have a couple wrapper layers but you wouldn't need to reimplement the handshake.

You'd end up with two copies of the WebSockets middleware (because SignalR also puts it in the pipeline): `WebSockets -> InterruptibleWebSockets -> WebSockets` but the WebSockets middleware is idempotent and intentionally **no-ops** if there's already an `IHttpWebSocketFeature` in place.